### PR TITLE
feat(compaction): support merge compaction group

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

part of https://github.com/risingwavelabs/risingwave/pull/17898, This PR introduces a new ability to merge existing Compaction Groups by adding a new GroupMerge delta, some changes introduced in this pr:

1. ctl support merge group
2. support GroupMerge for hummock
3. reorg code 
4. rename `GroupSplit` timer event to `GroupSchedule`

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by Li0k** [comment]
@wenym1 
hi. After merging this PR, we will have the ability to `Merge`, but this PR does not support auto-scheduling yet. We can explore how to proactively merge newly generated Groups through the `Merge` interface after Backfill.

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        }
+
+        let mut left_group_id = group_1;
+        let mut right_group_id = group_2;
+        let combine_member_table_ids = if combine_1.is_sorted() {
+            combine_1
+        } else {
+            std::mem::swap(&mut left_group_id, &mut right_group_id);
+            combine_2
+        };
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:102]: nits: it is unnecessary to have three `is_sorted()` check here given that `compaction_group_member_table_ids` returns back a BTreeSet. We can make the logic simpler yet slightly more efficient:
```rust
        debug_assert!(!member_table_ids_1.is_empty());
        debug_assert!(!member_table_ids_2.is_empty());
        debug_assert!(member_table_ids_1.is_sorted());
        debug_assert!(member_table_ids_2.is_sorted());
        
        // Make sure `member_table_ids_1` is smaller than `member_table_ids_2`
        let (left_group_id, right_group_id) =
            if member_table_ids_1.first().unwrap() < member_table_ids_2.first().unwrap() {
                (group_1, group_2)
            } else {
                std::mem::swap(&mut member_table_ids_1, &mut member_table_ids_2);
                (group_2, group_1)
            };

        // We can only merge two groups with non-overlapping member table ids
        if member_table_ids_1.last().unwrap() >= member_table_ids_2.first().unwrap() {
            return Err(Error::CompactionGroup(format!(
                "invalid merge group_1 {} group_2 {}",
                left_group_id, right_group_id
            )));
        }

        let combined_memtable_table_ids =
            member_table_ids_1.iter().chain(member_table_ids_2.iter());

        debug_assert!(combined_memtable_table_ids.is_sorted());
```

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+                .await
+                .try_get_compaction_group_config(group_1)
+                .unwrap()
+                .compaction_config()
+                .deref()
+                .clone();
+
+            {
+                // update config
+                config.split_weight_by_vnode = 0;
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:125]: Please document why we need to set this.

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+                    &self.metrics,
+                    right_group_id,
+                    right_group_max_level,
+                );
+            }
+
+            new_version_delta.pre_apply();
+
+            // purge right_group_id
+            compaction_groups_txn.purge(HashSet::from_iter(get_compaction_group_ids(
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:187]: Can we directly call `compaction_groups_txn.remove(...)`?

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs [hunk 4]
```diff
@@ -47,13 +48,17 @@ pub struct GroupDeltasSummary {
...
     pub group_meta_changes: Vec<GroupMetaChange>,
     pub group_table_change: Option<GroupTableChange>,
     pub new_vnode_partition_count: u32,
+    pub group_merge: Option<GroupMerge>,
 }
 
-pub fn summarize_group_deltas(group_deltas: &GroupDeltas) -> GroupDeltasSummary {
+pub fn summarize_group_deltas(
+    group_deltas: &GroupDeltas,
+    default_group_id: CompactionGroupId,
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:60]: what does the `default_group_id` mean?

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs [hunk 5]
```diff
@@ -835,6 +836,45 @@ impl HummockVersion {
...
+        left_group_id: CompactionGroupId,
+        right_group_id: CompactionGroupId,
+    ) {
+        // Double check
+        let left_group_id_table_ids = self
+            .state_table_info
+            .compaction_group_member_table_ids(left_group_id)
+            .iter()
+            .map(|table_id| table_id.table_id)
+            .collect_vec();
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:851]: No need to `collect_vec`

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs [hunk 6]
```diff
@@ -835,6 +836,45 @@ impl HummockVersion {
...
+
+        assert!(left_group_id_table_ids.is_sorted());
+        let right_group_id_table_ids = self
+            .state_table_info
+            .compaction_group_member_table_ids(right_group_id)
+            .iter()
+            .map(|table_id| table_id.table_id)
+            .collect_vec();
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:859]: No need to `collect_vec`

File: src/storage/hummock_sdk/src/compaction_group/mod.rs [hunk 7]
```diff
@@ -43,3 +43,105 @@ impl From<StaticCompactionGroupId> for CompactionGroupId {
...
+
+    use super::hummock_version_ext::{add_ssts_to_sub_level, insert_new_sub_level};
+    use crate::can_concat;
+    use crate::level::{Level, Levels};
+
+    pub fn merge_levels(left_levels: &mut Levels, right_levels: &mut Levels) {
+        let right_l0 = &right_levels.l0;
+
+        for right_sub_level in &right_l0.sub_levels {
+            let sub_level = right_sub_level.clone();
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/mod.rs:58]: IIUC, right_levels should be cleared eventually. Can we avoid `clone()` here but move?

File: src/storage/hummock_sdk/src/compaction_group/mod.rs [hunk 8]
```diff
@@ -43,3 +43,105 @@ impl From<StaticCompactionGroupId> for CompactionGroupId {
...
+            match get_sub_level_insert_hint(&left_levels.l0.sub_levels, &sub_level) {
+                Ok(insert_hint) => {
+                    add_ssts_to_sub_level(
+                        &mut left_levels.l0,
+                        insert_hint,
+                        sub_level.table_infos.clone(),
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/mod.rs:64]: ditto

File: src/storage/hummock_sdk/src/compaction_group/mod.rs [hunk 9]
```diff
@@ -43,3 +43,105 @@ impl From<StaticCompactionGroupId> for CompactionGroupId {
...
+                    );
+                }
+                Err(insert_hint) => {
+                    insert_new_sub_level(
+                        &mut left_levels.l0,
+                        sub_level.sub_level_id,
+                        sub_level.level_type,
+                        sub_level.table_infos.clone(),
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/mod.rs:72]: ditto

File: src/storage/hummock_sdk/src/compaction_group/mod.rs [hunk 10]
```diff
@@ -43,3 +43,105 @@ impl From<StaticCompactionGroupId> for CompactionGroupId {
...
+                        Some(insert_hint),
+                    );
+                }
+            }
+        }
+
+        left_levels
+            .l0
+            .sub_levels
+            .sort_by_key(|sub_level| sub_level.sub_level_id);
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/mod.rs:82]: Is this esstentially noop? Should we do an assertion here instead?

File: src/storage/hummock_sdk/src/compaction_group/mod.rs [hunk 11]
```diff
@@ -43,3 +43,105 @@ impl From<StaticCompactionGroupId> for CompactionGroupId {
...
+
+        // Reinitialise `vnode_partition_count`` to avoid misaligned hierarchies
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/mod.rs:84]: nits: extra ` here

File: src/storage/hummock_sdk/src/compaction_group/mod.rs [hunk 12]
```diff
@@ -43,3 +43,105 @@ impl From<StaticCompactionGroupId> for CompactionGroupId {
...
+            .sub_levels
+            .iter_mut()
+            .for_each(|sub_level| sub_level.vnode_partition_count = 0);
+
+        for (idx, level) in right_levels.levels.iter_mut().enumerate() {
+            if level.table_infos.is_empty() {
+                continue;
+            }
+
+            let insert_table_infos = level.table_infos.clone();
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/compaction_group/mod.rs:97]: ditto. Can you avoid `clone()` here?

**Review by hzxa21 — COMMENTED** [review]
Most comments are minor. Thanks for the efforts!

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs (hunk 2 above)
Comment (Li0k) [reply src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:125]: typo: remove it

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs (hunk 3 above)
Comment (Li0k) [reply src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:187]: Good Catch, We've already triggered Group Destory in the Merge Delta process, so we really don't need to purge again here.

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs (hunk 4 above)
Comment (Li0k) [reply src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:60]: It represents the group_id to which the delta version belongs, just for group destory

File: src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs (hunk 6 above)
Comment (Li0k) [reply src/storage/hummock_sdk/src/compaction_group/hummock_version_ext.rs:859]: ditto

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+            return Err(Error::CompactionGroup(format!(
+                "invalid merge group_1 {} group_2 {}",
+                left_group_id, right_group_id
+            )));
+        }
+
+        let combined_member_table_ids = member_table_ids_1
+            .iter()
+            .chain(member_table_ids_2.iter())
+            .collect_vec();
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:105]: nits: `collect_vec` is unnecessary

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs (hunk 3 above)
Comment (hzxa21) [reply src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs]: Can you explain more on this? I think here we are manipulating the `CompactionGroup` metadata in `CompactionGroupManager`, not in `HummockVersion`. Why is this related to GroupDestroy delta?

**Review by hzxa21 — APPROVED** [review]
LGTM. Given that the `sub_level_id` assignment is changed to not rely on epoch in this PR, I suggest we run another round of testing before merging.

**Review by zwang28 — APPROVED** [review]
LGTM

**Comment by Li0k** [comment]
Triggering multiple Splits and Merges in a Backfill test

https://grafana.test.risingwave-cloud.xyz/d/EpkBw5W4k/risingwave-dev-dashboard?from=now-1h&orgId=1&to=now&var-Filters=table_type%7C%21%3D%7CTABLE&var-component=All&var-datasource=P2453400D1763B4D9&var-instance=benchmark-risingwave&var-namespace=li0k-backfill-split-merge&var-pod=All&var-table=All

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+        for sst_id in versioning
+            .current_version
+            .get_sst_ids_by_group_id(left_group_id)
+            .chain(
+                versioning
+                    .current_version
+                    .get_sst_ids_by_group_id(right_group_id),
+            )
+        {
+            if !sst_id_set.insert(sst_id) {
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:119]: In which case will we have duplicate SST ids? Is it impossible and we just have the check here as a safeguard?

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+                return Err(Error::CompactionGroup(format!(
+                    "invalid merge group_1 {} group_2 {} duplicated sst_id {}",
+                    left_group_id, right_group_id, sst_id
+                )));
+            }
+        }
+
+        // TODO(li0k): remove this check (Since the current split_sst does not change key_range, this check can not be removed, otherwise concate will fail.)
+        // check branched sst on non-overlap level
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:128]: nits: this comment is stale?

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+            let right_levels = versioning
+                .current_version
+                .get_compaction_group_levels(group_2);
+
+            // we can not check the l0 sub level, because the sub level id will be rewritten when merge
+            // This check will ensure that other non-overlapping level ssts can be concat and that the key_range is correct.
+            let max_level = std::cmp::max(left_levels.levels.len(), right_levels.levels.len());
+            for level_idx in 0..max_level {
+                let left_level = left_levels.levels.get(level_idx).unwrap();
+                let right_level = right_levels.levels.get(level_idx).unwrap();
```
Comment (hzxa21) [inline src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:143]: It is confusing that `levels.get(level_idx)` is getting level `level_idx + 1`. At the first glance, I thought we are checking level 0 as well here. How about using the [`get_level`](https://github.com/risingwavelabs/risingwave/blob/4af2cd36dcdd3fcdd850a0c84a3bbce47e1c50ff/src/storage/hummock_sdk/src/level.rs#L214) here?

**Review by hzxa21 — COMMENTED** [review]
Rest LGTM

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs (hunk 14 above)
Comment (Li0k) [reply src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:119]: https://github.com/risingwavelabs/risingwave/pull/18431

Before 18431, commit_epoch may generate the duplicated sst_id, This check is just for forward compatibility.

File: src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs (hunk 15 above)
Comment (Li0k) [reply src/meta/src/hummock/manager/compaction/compaction_group_schedule.rs:128]: Indeed, we can always maintain this check

**Pull request merged**